src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,784 @@
+            m_renderer_filter = &(frame.get_filter());
+
+            // Fetch the tile bounds (inclusive).
+            m_tile_origin_x = static_cast<int>(tile_x * props.m_tile_width);
+            m_tile_origin_y = static_cast<int>(tile_y * props.m_tile_height);
+            m_tile_end_x = static_cast<int>(m_tile_origin_x + tile.get_width() - 1);
+            m_tile_end_y = static_cast<int>(m_tile_origin_y + tile.get_height() - 1);
+
+            for (int ry = m_tile_origin_y; ry <= m_tile_end_y; ++ry)
+            {
+                for (int rx = m_tile_origin_x; rx <= m_tile_end_x; ++rx)
+                {
+                    m_pixel_samples[ry * m_frame_width + rx] = WeightMap();
+                }
+            }
+
+            if (frame.has_crop_window())
+                m_crop_window = frame.get_crop_window();
+            else
+                m_crop_window = AABB2u(Vector2u(m_tile_origin_x, m_tile_origin_y), Vector2u(m_tile_end_x, m_tile_end_y));
+        }
+
+        void on_tile_end(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y) override
+        {
+            for (int ry = m_tile_origin_y; ry <= m_tile_end_y; ++ry)
+            {
+                for (int rx = m_tile_origin_x; rx <= m_tile_end_x; ++rx)
+                {
+                    const WeightMap& weight_map = m_pixel_samples[ry * m_frame_width + rx];
+
+                    if (weight_map.size() > 0)
+                    {
+                        float total_weight = 0.0f;
+                        for (const auto& item : weight_map)
+                            total_weight += item.second;
+
+                        if (total_weight == 0.0f)
+                            total_weight = 1.0f;
+
+                        vector<pair<float, uint32_t>> ranked_vector;
+                        for (const auto& item : weight_map)
+                            ranked_vector.push_back(make_pair(item.second, item.first));
+                        sort(ranked_vector.begin(), ranked_vector.end(),
+                            [](pair<float, uint32_t> const& a, pair<float, uint32_t> const& b) { return a.first > b.first; });
+
+                        uint32_t m3hash = ranked_vector[0].second;
+
+                        vector<float> pixel_values;
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : hash_to_float(m3hash));
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : ((float)((m3hash << 8)) / (float)UINT32_MAX));
+                        pixel_values.push_back(m3hash == 0 ? 0.0f : ((float)((m3hash << 16)) / (float)UINT32_MAX));
+
+                        // Remove background contribution
+                        if (ranked_vector.size() > 1 && m3hash == 0)
+                            ranked_vector.erase(ranked_vector.begin());
+
+                        for (size_t i = 0, e = ranked_vector.size(); i < e && i < m_num_layers; ++i)
```

Comment:
Set `e` to `min(ranked_vector.size(), m_num_layers)`?

Reply:
Thanks for this tip.